File: java-advent-of-code/src/main/java/com/mycompany/advent_of_code_java/Advent_of_code_2025_003_001.java
package com.mycompany.advent_of_code_java;

import module java.base;

public class Advent_of_code_2025_003_001 {

    public static void main(String[] args) {
        String Input_path = (args.length >= 1)
                ? args[0]
                : "E:\\documents-container\\advent-of-code\\java-advent-of-code\\inputs\\2025-003\\input.txt";

        try {
            BigInteger Total_output_joltage = Solve(Input_path);
            System.out.println(Total_output_joltage);
        } catch (IOException e) {
            System.err.println("Failed to read input: " + Input_path);
            e.printStackTrace(System.err);
            System.exit(1);
        } catch (IllegalArgumentException e) {
            System.err.println("Invalid input: " + e.getMessage());
            System.exit(2);
        }
    }

    public static BigInteger Solve(String Input_path) throws IOException {
        BigInteger Sum = BigInteger.ZERO;

        try (BufferedReader Buffered_reader = Files.newBufferedReader(Path.of(Input_path))) {
            String Line;
            long Line_number = 0;

            while ((Line = Buffered_reader.readLine()) != null) {
                Line_number++;
                Line = Line.trim();
                if (Line.isEmpty()) {
                    continue;
                }

                if (Line.length() < 2) {
                    throw new IllegalArgumentException("Line " + Line_number + " has fewer than 2 digits.");
                }

                int Max_joltage_for_bank = Max_two_digit_from_line(Line);
                Sum = Sum.add(BigInteger.valueOf(Max_joltage_for_bank));
            }
        }

        return Sum;
    }

    private static int Max_two_digit_from_line(String Bank_digits) {
        int Length = Bank_digits.length();

        int[] Digits = new int[Length];
        for (int i = 0; i < Length; i++) {
            char c = Bank_digits.charAt(i);
            if (c < '0' || c > '9') {
                throw new IllegalArgumentException("Non-digit character '" + c + "' in line: " + Bank_digits);
            }
            Digits[i] = c - '0';
        }

        int[] Suffix_max = new int[Length];
        Suffix_max[Length - 1] = Digits[Length - 1];
        for (int i = Length - 2; i >= 0; i--) {
            Suffix_max[i] = Math.max(Digits[i], Suffix_max[i + 1]);
        }

        int Best = -1;
        for (int i = 0; i < Length - 1; i++) {
            int Tens_digit = Digits[i];
            int Ones_digit = Suffix_max[i + 1];
            int Value = 10 * Tens_digit + Ones_digit;

            if (Value > Best) {
                Best = Value;
            }

            if (Best == 99) {
                break;
            }
        }

        return Best;
    }
}

File: python/Advent_of_code_2025_003_001.py
import sys
from pathlib import Path

def max_two_digit_from_line(bank_digits: str) -> int:
    """
    Finds the max 10*digits[i] + digits[j] where i < j.
    Uses a suffix max array to find the best 'ones' digit for any 'tens' digit.
    """
    length = len(bank_digits)
    
    # Convert string to list of integers and validate
    try:
        digits = [int(d) for d in bank_digits]
    except ValueError:
        raise ValueError(f"Non-digit character in line: {bank_digits}")

    # Build Suffix Max array: Suffix_max[i] is the max digit from index i to the end
    suffix_max = [0] * length
    suffix_max[-1] = digits[-1]
    for i in range(length - 2, -1, -1):
        suffix_max[i] = max(digits[i], suffix_max[i+1])

    best = -1
    for i in range(length - 1):
        tens_digit = digits[i]
        ones_digit = suffix_max[i + 1]
        value = 10 * tens_digit + ones_digit
        
        if value > best:
            best = value
            
        if best == 99: # Early exit optimization
            break
            
    return best

def solve(input_path_str: str) -> int:
    total_sum = 0
    input_path = Path(input_path_str)
    
    if not input_path.exists():
        raise FileNotFoundError(f"File not found: {input_path}")

    with open(input_path, 'r', encoding='utf-8') as f:
        for line_no, line in enumerate(f, 1):
            line = line.strip()
            if not line:
                continue
                
            if len(line) < 2:
                raise ValueError(f"Line {line_no} has fewer than 2 digits.")
                
            max_joltage = max_two_digit_from_line(line)
            total_sum += max_joltage
            
    return total_sum

def main():
    # Default path based on your environment
    default_path = r"E:\documents-container\advent-of-code\java-advent-of-code\inputs\2025-003\input.txt"
    
    input_path = sys.argv[1] if len(sys.argv) >= 2 else default_path
    
    try:
        ans = solve(input_path)
        print(ans)
    except FileNotFoundError:
        print(f"Failed to read input: {input_path}", file=sys.stderr)
        sys.exit(1)
    except ValueError as e:
        print(f"Invalid input: {e}", file=sys.stderr)
        sys.exit(2)

if __name__ == "__main__":
    main()